crates/voicevox_core/src/user_dict/word.rs

```diff
@@ -0,0 +1,225 @@
+            return Err(Error::InvalidWord(format!(
+                "優先度は{}以上{}以下である必要があります。",
+                MIN_PRIORITY, MAX_PRIORITY
+            )));
+        }
+        Self::validate_pronunciation(&pronunciation[..])?;
+        let mora_count = Self::calculate_mora_count(&pronunciation[..], accent_type)?;
+        Ok(Self {
+            surface: Self::to_zenkaku(&surface[..]),
+            pronunciation,
+            accent_type,
+            word_type,
+            priority,
+            mora_count,
+        })
+    }
+
+    /// カタカナの文字列が発音として有効かどうかを判定する。
+    fn validate_pronunciation(pronunciation: &str) -> Result<()> {
+        // 元実装：https://github.com/VOICEVOX/voicevox_engine/blob/39747666aa0895699e188f3fd03a0f448c9cf746/voicevox_engine/model.py#L190-L210
+        if !PRONUNCIATION_REGEX.is_match(pronunciation) {
+            return Err(Error::InvalidWord(
+                "発音は有効なカタカナである必要があります。".to_string(),
+            ));
+        }
+        let sutegana = ['ァ', 'ィ', 'ゥ', 'ェ', 'ォ', 'ャ', 'ュ', 'ョ', 'ヮ', 'ッ'];
+
+        let pronunciation_chars = pronunciation.chars().collect::<Vec<_>>();
+
+        for i in 0..pronunciation_chars.len() {
+            // 「キャット」のように、捨て仮名が連続する可能性が考えられるので、
+            // 「ッ」に関しては「ッ」そのものが連続している場合と、「ッ」の後にほかの捨て仮名が連続する場合のみ無効とする
+            if sutegana.contains(pronunciation_chars.get(i).unwrap())
+                && i < pronunciation_chars.len() - 1
+                && (sutegana[..sutegana.len() - 1]
+                    .contains(pronunciation_chars.get(i + 1).unwrap())
+                    || (pronunciation_chars.get(i).unwrap() == &'ッ'
+                        && sutegana.contains(pronunciation_chars.get(i + 1).unwrap())))
+            {
+                return Err(Error::InvalidWord(
+                    "無効な発音です。(捨て仮名の連続)".to_string(),
+                ));
+            }
+
+            if pronunciation_chars.get(i).unwrap() == &'ヮ'
+                && i != 0
+                && !['ク', 'グ'].contains(pronunciation_chars.get(i - 1).unwrap())
+            {
+                return Err(Error::InvalidWord(
+                    "無効な発音です。(「くゎ」「ぐゎ」以外の「ゎ」の使用)".to_string(),
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    /// カタカナの発音からモーラ数を計算する。
+    fn calculate_mora_count(pronunciation: &str, accent_type: usize) -> Result<usize> {
+        // 元実装：https://github.com/VOICEVOX/voicevox_engine/blob/39747666aa0895699e188f3fd03a0f448c9cf746/voicevox_engine/model.py#L212-L236
+        let rule_others =
+            r#"[イ][ェ]|[ヴ][ャュョ]|[トド][ゥ]|[テデ][ィャュョ]|[デ][ェ]|[クグ][ヮ]"#;
+        let rule_line_i = r#"[キシチニヒミリギジビピ][ェャュョ]"#;
+        let rule_line_u = r#"[ツフヴ][ァ]|[ウスツフヴズ][ィ]|[ウツフヴ][ェォ]"#;
+        let rule_one_mora = r#"[ァ-ヴー]"#;
+
+        let mora_count = regex::Regex::new(&format!(
+            r#"(?:{}|{}|{}|{})"#,
+            rule_others, rule_line_i, rule_line_u, rule_one_mora
+        ))
+        .unwrap()
+        .find_iter(pronunciation)
+        .count();
```

Comment:
ここは何度も通るので、`once_cell::sync::Lazy`で`static`にするのがいいかと思います。

あと`format!`ではなく文字列リテラルの`concat!`で正規表現を組み立てれば、[`clippy::invalid_regex`](https://rust-lang.github.io/rust-clippy/rust-1.70.0/index.html#invalid_regex)のサポートが受けられます。

```suggestion
        let mora_count = {
            static REGEX: Lazy<Regex> = Lazy::new(|| {
                Regex::new(concat!(
                    "(?:",
                    "[イ][ェ]|[ヴ][ャュョ]|[トド][ゥ]|[テデ][ィャュョ]|[デ][ェ]|[クグ][ヮ]|", // rule_others
                    "[キシチニヒミリギジビピ][ェャュョ]|", // rule_line_i
                    "[ツフヴ][ァ]|[ウスツフヴズ][ィ]|[ウツフヴ][ェォ]|", // rule_line_u
                    "[ァ-ヴー]",                           // rule_one_mora
                    ")",
                ))
                .unwrap()
            });
            &REGEX
        }
        .find_iter(pronunciation)
        .count();
```